lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/avro/ProtocolSerializerTest.java

```diff
@@ -0,0 +1,187 @@
+    sender2.onNext(serializer.write(new AvroTestMessage(numbers[1], strings[1]), 2));
+
+    final AvroMessageObserver avroObserver1 = new AvroMessageObserver();
+    final AvroMessageObserver avroObserver2 = new AvroMessageObserver();
+
+    serializer.read(queue1.take(), avroObserver1);
+    serializer.read(queue2.take(), avroObserver2);
+
+    assertEquals(numbers[0], avroObserver2.getNumber());
+    assertEquals(strings[0], avroObserver2.getDataString());
+
+    assertEquals(numbers[1], avroObserver1.getNumber());
+    assertEquals(strings[1], avroObserver1.getDataString());
+  }
+
+  /**
+   * Build a remote manager on the local IP address with an unused port.
+   * @param identifier The identifier of the remote manager.
+   * @return A RemoteManager instance listing on the local IP address
+   *         with a unique port number.
+   */
+  private RemoteManager getTestRemoteManager(final String identifier) {
+    RemoteManager remoteManager = null;
+    try {
+      int port = 0;
+      boolean order = true;
+      int retries = 3;
+      int timeOut = 10000;
+
+      final Injector injector = Tang.Factory.getTang().newInjector();
+      final LocalAddressProvider localAddressProvider = injector.getInstance(LocalAddressProvider.class);
+      final TcpPortProvider tcpPortProvider = injector.getInstance(TcpPortProvider.class);
+      final RemoteManagerFactory remoteManagerFactory = injector.getInstance(RemoteManagerFactory.class);
+
+      remoteManager = remoteManagerFactory.getInstance(
+        identifier, localAddressProvider.getLocalAddress(), port, new ByteCodec(),
+        new LoggingEventHandler<Throwable>(), order, retries, timeOut,
+        localAddressProvider, tcpPortProvider);
+
+    } catch (final Exception e) {
+      e.printStackTrace();
+      LOG.log(Level.SEVERE, "Initialization failed: " + e.getMessage());
+    }
+    return remoteManager;
+  }
+
+  private final class ByteMessageObserver implements EventHandler<RemoteMessage<byte[]>> {
+    private BlockingQueue<byte[]> queue;
+
+    /**
+     * @param queue Queue where incoming messages will be stored.
+     */
+    ByteMessageObserver(final BlockingQueue<byte[]> queue) {
+      this.queue = queue;
+    }
+
+    /**
+     * Deserialize and direct incoming messages to the registered MuiltiObserver event handler.
+     * @param message A RemoteMessage<byte[]> object which will be deserialized.
+     */
+    public void onNext(final RemoteMessage<byte[]> message) {
+      queue.add(message.getMessage());
+    }
+  }
+
+  /**
+   * Processes messages from the network remote manager.
+   */
+  public final class AvroMessageObserver extends MultiObserverImpl<AvroMessageObserver> {
+    private int number;
+    private String dataString;
+
+    int getNumber() {
+      return number;
+    }
+
+    String getDataString() {
+      return dataString;
+    }
+
+    public void onError(final Exception error) {
+      LOG.log(Level.SEVERE, "OnError: ", error.getMessage());
```

Comment:
Also, do we still have to have this method? You've removed it from the `MultiObserver` interface